src/System.Linq/src/System/Linq/Concat.cs

```diff
@@ -61,16 +60,15 @@ internal Concat2Iterator(IEnumerable<TSource> first, IEnumerable<TSource> second
 
             internal override ConcatIterator<TSource> Concat(IEnumerable<TSource> next)
             {
-                bool hasOnlyCollections = _first is ICollection<TSource> &&
-                                          _second is ICollection<TSource> &&
-                                          next is ICollection<TSource>;
+                bool hasOnlyCollections = next is ICollection<TSource> &&
+                                          _first is ICollection<TSource> &&
+                                          _second is ICollection<TSource>;
```

Comment:
TBH, I thought I'd undone that as not relevant to what's being done here, but clearly I'd committed it. It seems reasonable that if we're going to do something on two fields and a local and any of them can short-circuit away from the rest to operate on the local first.